### ISSUE TITLE
dnsdist: incoming PROXY protocol as option to addLocal

### ISSUE THREAD

**Issue by wojas** [PowerDNS/pdns#12922]
- Program: dnsdist
 - Issue type: Feature request

### Short description

Support enabling incoming PROXY protocol support on a per-port basis using a new `addLocal` option.

### Usecase

- Debugging filtering in a containerized environment
- Debugging filtering without losing the ability to do normal DNS queries without PROXY protocol from that address
- Support a mixed environment where the IPs of a load balancer with PROXY support are not known in advance, and normal DNS queries also need to be supported.

### Description

Currently `setProxyProtocolACL` can be used to enable PROXY protocol for incoming queries for certain addresses. When enabled, the PROXY protocol is also required from these configured sources.

This complicates debugging in development and production deployments that use containers. Overriding the source IP with the PROXY protocol can be very useful to test filtering when a query comes from specific source IPs (e.g. using `kdig +proxy=0.1.2.3-0.0.0.0`), but if you enable it for all IPs in a container, you can no longer perform normal queries without PROXY header against it.

If it were possible to enable the PROXY protocol on a per-port basis, every deployment could expose an additional port with the PROXY protocol enabled for debugging, that is only accessible from trusted IPs.

Other servers like nginx make this configurable on a per-port basis: `listen 12345 proxy_protocol;`.

Example:

```lua
addLocal('0.0.0.0:53')
addLocal('192.0.2.1:5300', {proxyProtocol=true})
```

### Alternative

Alternatively, dnsdist could allow ordinary queries without the PROXY protocol headers on `setProxyProtocolACL` IPs.

This would solve debugging issues for development environment, but have reduced utility in production environments and result in a higher risk of security issues due to misconfiguration.

**Comment by franklouwers** [comment]
Thanks Remi. Any chance we’d see this for rec/auth one day?On 8 Dec 2023, at 17:14, Remi Gacogne ***@***.***> wrote:﻿
Closed #12922 as completed via #13517.

**Comment by omoerbeek** [comment]
> Thanks Remi. Any chance we’d see this for rec/auth one day?On 8 Dec 2023, at 17:14, Remi Gacogne ***@***.***> wrote:﻿ Closed #12922 as completed via #13517. —Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you commented.Message ID: ***@***.***>

Testing PowerGit's handling of funny chars

**Comment by rgacogne** [comment]
> Thanks Remi. Any chance we’d see this for rec/auth one day?

It can theoretically be done, of course, but the way listening IPs and ports are configured in the authoritative and recursor might make it quite complicated. I also don't know if Peter and/or Otto have plans to do that in the future.
